🚀 3단계 - 사다리(게임 실행)

ljg294 opened the pull request:
* refactor/feedback-step2
  * refactored codes based on feedback from step2
* feature/ladder-game
  * implement ladder game
    * add position field in player
    * modified codes related with ladder game
    * modified test codes as well
    
규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
-> 이 내용을 지키지 못한 점이 있습니다....
제가 잘못 접근한게 아닌가 싶네요... :(

catsbi commented on src/main/java/ladder/domain/Destination.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
IDE  자동생성 코드도 블럭`{}`생략은 지양해주세요!

catsbi commented on src/main/java/ladder/domain/Ladder.java:
```diff
@@ -10,31 +11,33 @@
+    public Ladder(int width,
                   int height){
-        this(createLadder(players, height));
+        this(createLadder(width, height));
```
처음에는 간단하게 생각해보면 좋을 것 같아요. 
| 생성자에서는 유효성 검증 외에 비즈니스 로직을 가지지 않는다. 

왜 그래야 할까요?

ljg294 replied:
SRP 때문인 것 같아 보이네요
생성자에서는 유효성 검증, 객체 초기화 외에 다른 역할을 가질 필요가 없다는것 인것 같네요
Ladder라는 객체읭 상태를 설정관련 로직만 가져야 한다 라고 해석할 수도 있겠네요

catsbi commented on src/main/java/ladder/domain/factory/PlayerFactory.java:
```diff
@@ -16,8 +18,28 @@ public static Players create(List<String> playersNames){
+                .mapToObj(index -> new Player(playersNames.get(index), index))
                 .collect(Collectors.toList());
+
+//        List<Player> playerList = new ArrayList<>();
+//        for (int index = 0; index < playersNames.size(); index++) {
+//            Player player = new Player(playersNames.get(index), index);
+//            playerList.add(player);
+//        }
+//
+//        return playerList;
+
+//        return playersNames.stream()
+//                .map(Player::new)
+//                .collect(Collectors.toList());
```
사용되지 않는 코드는 제거해주세요!

catsbi commented on src/main/java/ladder/domain/factory/PlayerFactory.java:
```diff
@@ -16,8 +18,28 @@ public static Players create(List<String> playersNames){
+    }
+
+    private static void validatePlayers(List<String> playersNames){
+        if (playersNames.size() < 2){
```
매직 넘버는 분리해주세요!

catsbi commented on src/main/java/ladder/service/LadderGame.java:
```diff
@@ -0,0 +1,75 @@
+
+public class LadderGame {
+
+    private final Players players;
+    private final Destinations destinations;
+    private final Ladder ladder;
```
서비스 계층은 기본적으로 무상태(stateless) 를 유지하도록 해주세요. 이러한 부분은 클린 아키텍처를 참고해보시길 바래요.  서비스 계층은 별도의 상태에 종속되는 개념이 아닙니다. 도메인이 특정 상태나 개념에 대한 책임을 진다면, 서비스 계층은 특정 비즈니스 로직에 대한 책임을 지고 각 도메인에게 요청/응답을 하고 이 행위들을 취합해서 완성된 비즈니스 로직의 결과를 반환하는게 책임이거든요. 즉, 별도의 상태를 가지고 의존성이 생길 필요가 없다는 의미이기도 합니다.

catsbi reviewed (changes requested):
재규님 3단계 미션 구현하신부분들 확인해봤습니다. 
코멘트 주신 부분들 중 인자가 3개가 넘는곳을 찾아봤는데, 일단 생성자는 3개라는 제약과 무관합니다. 물론, 3개 이상이면서 동일한 타입이 연달아 있는 생성자인경우 빌더패턴을 권장드리기는 합니다. 

그리고 그게 아닌 일반 함수에서 3개 이상이 필요하다면 해당 함수의 책임이 적절한지, 고민해보면 좋을 것 같아요. 

피드백드린 부분 고민해서 다시 리뷰요청 부탁드려요!

ljg294 commented:
* refactored codes based on feedback
    * separated with LadderBody
    * modified LadderGame as a service layer (made it stateless)
    * modified test codes as well

catsbi commented on src/main/java/ladder/domain/factory/LadderFactory.java:
```diff
@@ -0,0 +1,33 @@
+
+public class LadderFactory {
+
+    private LadderFactory() {
+    }
+
+    private static final int MIN_PLAYERS = 2;
```
보통 상수의 위치가 클래스의 최상단에 위치하는게 컨벤션이기는 합니다 ㅎㅎ

ljg294 replied:
엇 실수했네요 좀 더 꼼꼼하게 진행하겠습니다 :)

catsbi reviewed (approved):
재규님 3단계 피드백 반영까지 잘 해주셨어요 👍 
크게 문제 안보이는 것 같아 머지드립니다. 다음 미션 진행해주세요!
